ruby/sdk-lightweight.md

```diff
@@ -0,0 +1,314 @@
+block (that will not raise until the block has executed) or opt-out of this default behaviour
+altogether and use `activity.cancelled?` flag to determine when an activity was cancelled.
+
+Here is what it will look like from a workflow perspective:
+
+```ruby
+class MyWorkflow < Temporal::Activity
+  def execute
+    future = MyActivity.execute
+
+    workflow.sleep(10)
+
+    future.cancel!
+  end
+end
+```
+
+*NOTE: `future.cancel` returns a cancellation future that can be used as any other future.*
+
+
+## Execute a child workflow
+
+`parent_workflow.rb`
+
+```ruby
+class ChildWorkflow < Temporal::Workflow
+  def execute(name)
+    "Hello, #{name}!"
+  end
+end
+
+class ParentWorkflow < Temporal::Workflow
+  def execute
+    ChildWorkflow.execute!('parent', id: 'id-2')
+  end
+end
+```
+
+
+## Async workflow
+
+Calling `#execute` (rather than `#execute!`) on activities and workflows results in their
+non-blocking execution.
+
+```ruby
+require 'sum_activity'
+
+class AsyncWorkflow < Temporal::Workflow
+  def execute
+    activity_future = SumActivity.execute
+    workflow_future = ParentWorkflow.execute
+
+    workflow.wait_for(activity_future, workflow_future)
+
+    activity_future.result # => 3
+    workflow_future.result # => 'Hello, parent!'
+  end
+end
+```
+
+
+## Register signal and query handlers
+
+`calculator_workflow.rb`
+
+```ruby
+class CalculatorWorkflow < Temporal::Workflow
+  def execute
+    value = 0
+
+    workflow.on_signal('add') do |val|
+      value += val
+    end
+
+    workflow.on_signal('subtract') do |val|
+      value -= val
+    end
+
+    workflow.on_query('get_value') do
+      value
+    end
+
+    # A condition block will pause until the enclosed statement returns true
+    workflow.wait_for { value >= 42 }
+  end
+end
+```
+
+#### Alternative approach
+
+To provide extra flexibility in handling signals and queries we will allow to defined signal and
+query handlers as methods:
+
+```ruby
+class CalculatorWorkflow < Temporal::Workflow
+  # Both signal and query handlers can be defined as Symbols (when the name matches the
+  # method) or a Hash when name is different (or can not be used as a method name)
+  signals :add, 'my.signal.subtract' => :subtract
+  queries :get_value, 'my.query.value' => :get_value
+
+  # It is preferable in this case to initialize the values before the execution to
+  # support starting a workflow with a signal
+  def initialize
+    @value = 0
+  end
+
+  def execute
+    workflow.wait_for { value >= 42 }
+  end
+
+  private
+
+  def add(val)
+    @value += val
+  end
+
+  def subtract(val)
+    @value -= val
+  end
+
+  def get_value
+    @value
+  end
+end
+```
+
+
+## Signal and query a workflow
+
+```ruby
+require 'calculator_workflow'
+
+connection = Temporal::Connection.new('localhost:7233')
+client = Temporal::Client.new(connection, namespace: 'demo')
+
+handle = client.start_workflow(SumWorkflow.execute(1, 2), id: 'id-1', task_queue: 'demo')
+handle.signal('add', 34)
```

Comment:
You support `client.start_workflow(SumWorkflow.execute(1, 2), ...` so why not support `handle.signal(SumWorkflow.add(34))` here? (of course still supporting the string option for when you don't have the method handy)

Reply:
I definitely want to avoid things like underscore prefixes and pieces like that. Also remember we want these things to be typed some day in the future.